### PR TITLE
Split tool

### PR DESCRIPTION
**Pull request by 3nids**
This allows splitting features using a line.

The new tool is embed in a new "geometry editors toolbar".

The editors are hold in a cpp QStandardItemModel which is filled up on QML side.
Hence, it is possible to add new editors on QML side only.

Some changes coming with this PR:
* the digitizing toolbar has been simplified a bit, only using an option to show or not the confirm button (no confirm button for measure tool)
* logic (JS code) of digit toolbar moved its dedicated file
* close tool moved to dedicated QML file

To go further:
* add an option for topological mode?

[QField](https://www.youtube.com/watch?v=JyOx-EJEFSo)

### CODE REVIEW

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-698-b2e3f8464cc3e2a72f90f5b010198a960ad7bd3a-armv7.apk)
 - [arm64_v8a](https://download.opengis.ch/qfield/ci-builds/qfield-dev-698-b2e3f8464cc3e2a72f90f5b010198a960ad7bd3a-arm64_v8a.apk)
 - [x86](https://download.opengis.ch/qfield/ci-builds/qfield-dev-698-b2e3f8464cc3e2a72f90f5b010198a960ad7bd3a-x86.apk)
 - [x86_64](https://download.opengis.ch/qfield/ci-builds/qfield-dev-698-b2e3f8464cc3e2a72f90f5b010198a960ad7bd3a-x86_64.apk)

File: src/qgsquick/qgsquickutils.cpp [hunk 1]
```diff
@@ -323,3 +323,11 @@ qreal QgsQuickUtils::calculateScreenDensity()
   int dpi = dpiX < dpiY ? dpiX : dpiY; // In case of asymmetrical DPI. Improbable
   return dpi / 160.;  // 160 DPI is baseline for density-independent pixels in Android
 }
+
+void QgsQuickUtils::selectFeaturesInLayer( QgsVectorLayer *layer, const QList<int> &fids, QgsVectorLayer::SelectBehavior behavior )
```
Comment (3nids) [inline src/qgsquick/qgsquickutils.cpp]: corresponding PR in QGIS: https://github.com/qgis/QGIS/pull/32735

**Comment by m-kuhn** [comment]
@3nids does this require a CMakeLists update as well?

**Comment by 3nids** [comment]
@m-kuhn CMakeLists.txt updated

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-698-c5a43496fbd2fbe9e4c008c7bdc6d03cdf4a2d81-armv7.apk)
 - [arm64_v8a](https://download.opengis.ch/qfield/ci-builds/qfield-dev-698-c5a43496fbd2fbe9e4c008c7bdc6d03cdf4a2d81-arm64_v8a.apk)
 - [x86](https://download.opengis.ch/qfield/ci-builds/qfield-dev-698-c5a43496fbd2fbe9e4c008c7bdc6d03cdf4a2d81-x86.apk)
 - [x86_64](https://download.opengis.ch/qfield/ci-builds/qfield-dev-698-c5a43496fbd2fbe9e4c008c7bdc6d03cdf4a2d81-x86_64.apk)

File: src/core/geometryeditorsmodel.cpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+{
+  if ( mVertexModel == vertexModel )
+    return;
+  mVertexModel = vertexModel;
+  emit vertexModelChanged();
+}
+
+
+
+QVariant GeometryEditorsModel::data( const QModelIndex &index, int role ) const
```
Comment (m-kuhn) [inline src/core/geometryeditorsmodel.cpp]: Does this add anything to the parent method?

File: src/core/maptoolmodel.cpp [hunk 3]
```diff
@@ -0,0 +1,6 @@
+#include "maptoolmodel.h"
```
Comment (m-kuhn) [inline src/core/maptoolmodel.cpp]: What's this file for?

File: src/core/maptoolmodel.h [hunk 4]
```diff
@@ -0,0 +1,13 @@
+#ifndef MAPTOOLMODEL_H
```
Comment (m-kuhn) [inline src/core/maptoolmodel.h]: What's this file for?

File: src/qml/CloseTool.qml [hunk 5]
```diff
@@ -0,0 +1,58 @@
+import QtQuick 2.12
+import QtQuick.Controls 2.12
+import Theme 1.0
+
+
+ToolButton {
+  id: closeTool
+
+  property string toolText: qStr("close")
```
Comment (m-kuhn) [inline src/qml/CloseTool.qml]:
```suggestion
  property string toolText: qsTr("close")
```

?

File: src/qml/CloseTool.qml [hunk 6]
```diff
@@ -0,0 +1,58 @@
+import QtQuick 2.12
+import QtQuick.Controls 2.12
+import Theme 1.0
+
+
+ToolButton {
```
Comment (m-kuhn) [inline src/qml/CloseTool.qml]: Can you add a docstring, what a CloseTool is for?

File: src/core/qgismobileapp.h [hunk 7]
```diff
@@ -117,6 +118,14 @@ class QgisMobileapp : public QQmlApplicationEngine
       return singletonClass;
     }
 
+    static QObject *geometryEditorsSingletonProvider( QQmlEngine *engine, QJSEngine *scriptEngine )
+    {
+      Q_UNUSED( engine )
```
Comment (m-kuhn) [inline src/core/qgismobileapp.h]: Code to cpp?

Comment (3nids) [reply src/core/qgismobileapp.h]: What do you mean?

File: src/core/maptoolmodel.cpp (hunk 3 above)
Comment (3nids) [reply src/core/maptoolmodel.cpp]: To test the reviewer ;)
I will drop it

File: src/core/qgismobileapp.h (hunk 7 above)
Comment (m-kuhn) [reply src/core/qgismobileapp.h]: Move the definition from the header to the cpp file?

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-698-e1c572d052b44a72a0c95cb4ae2990bd65e0a512-armv7.apk)
 - [arm64_v8a](https://download.opengis.ch/qfield/ci-builds/qfield-dev-698-e1c572d052b44a72a0c95cb4ae2990bd65e0a512-arm64_v8a.apk)
 - [x86](https://download.opengis.ch/qfield/ci-builds/qfield-dev-698-e1c572d052b44a72a0c95cb4ae2990bd65e0a512-x86.apk)
 - [x86_64](https://download.opengis.ch/qfield/ci-builds/qfield-dev-698-e1c572d052b44a72a0c95cb4ae2990bd65e0a512-x86_64.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-698-965379eddbadb87cbb1ab8ee6afe24d12536e9a8-armv7.apk)
 - [arm64_v8a](https://download.opengis.ch/qfield/ci-builds/qfield-dev-698-965379eddbadb87cbb1ab8ee6afe24d12536e9a8-arm64_v8a.apk)
 - [x86](https://download.opengis.ch/qfield/ci-builds/qfield-dev-698-965379eddbadb87cbb1ab8ee6afe24d12536e9a8-x86.apk)
 - [x86_64](https://download.opengis.ch/qfield/ci-builds/qfield-dev-698-965379eddbadb87cbb1ab8ee6afe24d12536e9a8-x86_64.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-698-6829bebaa00a60b3966eac7aab8db8c6bf141e89-armv7.apk)
 - [arm64_v8a](https://download.opengis.ch/qfield/ci-builds/qfield-dev-698-6829bebaa00a60b3966eac7aab8db8c6bf141e89-arm64_v8a.apk)
 - [x86](https://download.opengis.ch/qfield/ci-builds/qfield-dev-698-6829bebaa00a60b3966eac7aab8db8c6bf141e89-x86.apk)
 - [x86_64](https://download.opengis.ch/qfield/ci-builds/qfield-dev-698-6829bebaa00a60b3966eac7aab8db8c6bf141e89-x86_64.apk)

**Comment by 3nids** [comment]
with the latest version, you cannot switch tools while modifying the geometry

**Pull request merged**